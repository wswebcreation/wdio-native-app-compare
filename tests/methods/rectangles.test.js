import { IMAGE_STRING } from '../mocks/mocks'

describe('rectangles', () => {
    let getDeviceInfoSpy, ignoreRectanglesOptions, Rectangles, DeviceInfo

    const deviceInfo = {
        dpr: 2,
        isSmallIphone: true,
        isLargeIphone: false,
        isMediumIphone: false,
        isExtraLargeIphone: false,
        screenSize: {
            height: 1700,
            width: 812,
        },
        screenshotHeight: 1700,
        screenshotWidth: 812,
        isPortrait: true,
        rectangles: {
            androidNavigationBar: {
                bottom: 800,
                top: 760,
                left: 0,
                right: 400,
            },
            iOSHomeBar: {
                bottom: 5,
                top: 883,
                left: 133,
                right: 148,
            },
            statusBar: {
                bottom: 26,
                top: 0,
                left: 0,
                right: 812,
            },
        },
    }

    beforeEach(() => {
        jest.isolateModules(() => {
            Rectangles = require('../../lib/methods/rectangles')
            DeviceInfo = require('../../lib/methods/getDeviceInfo')
        })
        delete global.driver
        delete global.$$

        global.$$ = () => [{ 'element-selector-1': {} }]
        global.driver = {
            capabilities: {},
            getElementRect: jest
                .fn()
                .mockResolvedValue({ x: 1, y: 2, width: 10, height: 20 }),
            getSettings: jest
                .fn()
                .mockResolvedValue({ allowInvisibleElements: false }),
            updateSettings: jest.fn().mockResolvedValue({}),
            isIOS: false,
            isAndroid: false,
        }
        getDeviceInfoSpy = jest
            .spyOn(DeviceInfo, 'getDeviceInfo')
            .mockResolvedValue(deviceInfo)
        ignoreRectanglesOptions = {
            blockOuts: [],
            blockOutNavigationBar: false,
            blockOutStatusBar: false,
            elementBlockOuts: [],
        }
    })

    afterEach(() => {
        global.driver = {
            getElementRect: jest.fn().mockRestore(),
            getSettings: jest.fn().mockRestore(),
            updateSettings: jest.fn().mockRestore(),
        }
        getDeviceInfoSpy.mockRestore()
    })

    describe('determineIgnoreRectangles', () => {
        it('should return nothing if no ignore rectangles need to be returned', async () => {
            expect(
                await Rectangles.determineIgnoreRectangles(
                    IMAGE_STRING,
                    ignoreRectanglesOptions
                )
            ).toMatchSnapshot()
        })

        it('should get the status bar block outs', async () => {
            ignoreRectanglesOptions.blockOutStatusBar = true

            expect(
                await Rectangles.determineIgnoreRectangles(
                    IMAGE_STRING,
                    ignoreRectanglesOptions
                )
            ).toMatchSnapshot()
        })

        it('should get the android navigation bar block outs', async () => {
            global.driver.isAndroid = true
            ignoreRectanglesOptions.blockOutNavigationBar = true

            expect(
                await Rectangles.determineIgnoreRectangles(
                    IMAGE_STRING,
                    ignoreRectanglesOptions
                )
            ).toMatchSnapshot()
        })

        it('should get the block outs', async () => {
            ignoreRectanglesOptions.blockOuts = [
                { x: 11, y: 22, width: 33, height: 44 },
                {
                    x: 21,
                    y: 32,
                    width: 43,
                    height: 54,
                },
            ]

            expect(
                await Rectangles.determineIgnoreRectangles(
                    IMAGE_STRING,
                    ignoreRectanglesOptions
                )
            ).toMatchSnapshot()
        })

        it('should return iPhone X bottom bar rectangles', async () => {
            ignoreRectanglesOptions.blockOutIphoneHomeBar = true
            global.driver.isIOS = true

            expect(
                await Rectangles.determineIgnoreRectangles(
                    IMAGE_STRING,
                    ignoreRectanglesOptions
                )
            ).toMatchSnapshot()
        })

        it('should return iPhone 12 bottom bar rectangles', async () => {
            getDeviceInfoSpy.mockResolvedValue({
                ...deviceInfo,
                isMediumIphone: true,
            })
            ignoreRectanglesOptions.blockOutIphoneHomeBar = true
            global.driver.isIOS = true

            expect(
                await Rectangles.determineIgnoreRectangles(
                    IMAGE_STRING,
                    ignoreRectanglesOptions
                )
            ).toMatchSnapshot()
        })

        it('should return iPhone 11 bottom bar rectangles', async () => {
            getDeviceInfoSpy.mockResolvedValue({
                ...deviceInfo,
                isLargeIphone: true,
            })
            ignoreRectanglesOptions.blockOutIphoneHomeBar = true
            global.driver.isIOS = true

            expect(
                await Rectangles.determineIgnoreRectangles(
                    IMAGE_STRING,
                    ignoreRectanglesOptions
                )
            ).toMatchSnapshot()
        })

        it('should return iPhone 12 Pro Max bottom bar rectangles', async () => {
            getDeviceInfoSpy.mockResolvedValue({
                ...deviceInfo,
                isExtraLargeIphone: true,
            })
            ignoreRectanglesOptions.blockOutIphoneHomeBar = true
            global.driver.isIOS = true

            expect(
                await Rectangles.determineIgnoreRectangles(
                    IMAGE_STRING,
                    ignoreRectanglesOptions
                )
            ).toMatchSnapshot()
        })

        it('should call the settings API for android when elements needs to be blocked out', async () => {
            global.driver.isAndroid = true
            ignoreRectanglesOptions.elementBlockOuts = [
                { element: { selector: 'selector-1' } },
                {
                    element: { selector: 'selector-1' },
                },
            ]
            await Rectangles.determineIgnoreRectangles(
                IMAGE_STRING,
                ignoreRectanglesOptions
            )
            expect(global.driver.getSettings).toBeCalled()
            expect(global.driver.updateSettings).toBeCalledWith({
                allowInvisibleElements: true,
            })
            expect(global.driver.updateSettings).toHaveBeenLastCalledWith({
                allowInvisibleElements: false,
            })
        })
    })

    describe('determineValidBlockOuts', () => {
        it('should return the valid blockouts if an valid array if provided', () => {
            const blockOuts = [
                { x: 1, y: 2, width: 3, height: 4 },
                { x: 1, y: 2, width: 3, height: 4 },
            ]

            expect(
                Rectangles.determineValidBlockOuts(blockOuts)
            ).toMatchSnapshot()
        })
        it('should return the valid blockouts if an array with invalid keys is provided', () => {
            const blockOuts = [
                { x: 5, y: 6, width: 7, height: 8 },
                { b: 5, y: 6, widthy: 7, height: 8 },
                { x: 9, y: 12, width: 13, height: 14 },
                { x: 9, ho: 12, width: 13, bar: 14 },
            ]

            expect(
                Rectangles.determineValidBlockOuts(blockOuts)
            ).toMatchSnapshot()
        })
    })

    describe('determineElementBlockOuts', () => {
        it('should return an empty array if no elements are provided', async () => {
            expect(
                await Rectangles.determineElementBlockOuts({
                    components: [],
                    dpr: 1,
                })
            ).toMatchSnapshot()
        })

        it('should return an array of one element blockout if 1 element is provided', async () => {
            const selector = 'element-selector-1'
            const components = [{ element: { selector } }]
            global.$$ = () => [{ 'element-selector-1': {} }]

            expect(
                await Rectangles.determineElementBlockOuts({
                    components: components,
                    dpr: 1,
                })
            ).toMatchSnapshot()
        })

        it('should return an array of one element blockout if multiple elements are provided but 1 with an element number', async () => {
            const selector = 'element-selector-1'
            const components = [{ element: { selector }, elementNumber: 1 }]
            global.$$ = () => [
                { 'element-selector-1': {} },
                { 'element-selector-1': {} },
                { 'element-selector-1': {} },
            ]

            expect(
                await Rectangles.determineElementBlockOuts({
                    components: components,
                    dpr: 1,
                })
            ).toMatchSnapshot()
        })

        it('should return an array of element blockouts if multiple elements are provided', async () => {
            const selector = 'element-selector-1'
            const components = [
                { element: { selector } },
                { element: { selector } },
                { element: { selector } },
            ]
            global.$$ = () => [
                { 'element-selector-1': {} },
                { 'element-selector-1': {} },
                { 'element-selector-1': {} },
            ]

            expect(
                await Rectangles.determineElementBlockOuts({
                    components: components,
                    dpr: 1,
                })
            ).toMatchSnapshot()
        })
    })

    describe('determineDimensions', () => {
        it('should be able to determine the dimensions when no margin is provided', () => {
            expect(
                Rectangles.determineDimensions(
                    {
                        x: 100,
                        y: 200,
                        width: 1000,
                        height: 2000,
                    },
                    0
                )
            ).toMatchSnapshot()
        })

        it('should be able to determine the dimensions when a positive margin is provided', () => {
            expect(
                Rectangles.determineDimensions(
                    {
                        x: 100,
                        y: 200,
                        width: 1000,
                        height: 2000,
                    },
                    25
                )
            ).toMatchSnapshot()
        })

        it('should be able to determine the dimensions when a negative margin is provided', () => {
            expect(
                Rectangles.determineDimensions(
                    {
                        x: 100,
                        y: 200,
                        width: 1000,
                        height: 2000,
                    },
                    -25
                )
            ).toMatchSnapshot()
        })
    })

    describe('getElementRectangles', () => {
        it('should return element rectangles', async () => {
            expect(
                await Rectangles.getElementRectangles({ elementId: 1 })
            ).toMatchSnapshot()
            expect(global.driver.getElementRect).toBeCalledWith(1)
        })
    })
})
