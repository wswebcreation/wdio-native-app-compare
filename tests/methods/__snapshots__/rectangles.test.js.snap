// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rectangles determineDimensions should be able to determine the dimensions when a negative margin is provided 1`] = `
{
  "bottom": 1950,
  "left": 125,
  "right": 950,
  "top": 225,
}
`;

exports[`rectangles determineDimensions should be able to determine the dimensions when a positive margin is provided 1`] = `
{
  "bottom": 2050,
  "left": 75,
  "right": 1050,
  "top": 175,
}
`;

exports[`rectangles determineDimensions should be able to determine the dimensions when no margin is provided 1`] = `
{
  "bottom": 2000,
  "left": 100,
  "right": 1000,
  "top": 200,
}
`;

exports[`rectangles determineElementBlockOuts should return an array of element blockouts if multiple elements are provided 1`] = `
[
  {
    "bottom": 20,
    "left": 1,
    "right": 10,
    "top": 2,
  },
  {
    "bottom": 20,
    "left": 1,
    "right": 10,
    "top": 2,
  },
  {
    "bottom": 20,
    "left": 1,
    "right": 10,
    "top": 2,
  },
  {
    "bottom": 20,
    "left": 1,
    "right": 10,
    "top": 2,
  },
  {
    "bottom": 20,
    "left": 1,
    "right": 10,
    "top": 2,
  },
  {
    "bottom": 20,
    "left": 1,
    "right": 10,
    "top": 2,
  },
  {
    "bottom": 20,
    "left": 1,
    "right": 10,
    "top": 2,
  },
  {
    "bottom": 20,
    "left": 1,
    "right": 10,
    "top": 2,
  },
  {
    "bottom": 20,
    "left": 1,
    "right": 10,
    "top": 2,
  },
]
`;

exports[`rectangles determineElementBlockOuts should return an array of one element blockout if 1 element is provided 1`] = `
[
  {
    "bottom": 20,
    "left": 1,
    "right": 10,
    "top": 2,
  },
]
`;

exports[`rectangles determineElementBlockOuts should return an array of one element blockout if multiple elements are provided but 1 with an element number 1`] = `
[
  {
    "bottom": 20,
    "left": 1,
    "right": 10,
    "top": 2,
  },
]
`;

exports[`rectangles determineElementBlockOuts should return an empty array if no elements are provided 1`] = `[]`;

exports[`rectangles determineIgnoreRectangles should get the android navigation bar block outs 1`] = `
[
  {
    "bottom": 800,
    "left": 0,
    "right": 400,
    "top": 760,
  },
]
`;

exports[`rectangles determineIgnoreRectangles should get the block outs 1`] = `
[
  {
    "bottom": 44,
    "left": 11,
    "right": 33,
    "top": 22,
  },
  {
    "bottom": 54,
    "left": 21,
    "right": 43,
    "top": 32,
  },
]
`;

exports[`rectangles determineIgnoreRectangles should get the status bar block outs 1`] = `
[
  {
    "bottom": 26,
    "left": 0,
    "right": 812,
    "top": 0,
  },
]
`;

exports[`rectangles determineIgnoreRectangles should return iPhone 11 bottom bar rectangles 1`] = `
[
  {
    "bottom": 5,
    "left": 133,
    "right": 148,
    "top": 883,
  },
  {
    "bottom": 5,
    "left": 133,
    "right": 148,
    "top": 883,
  },
]
`;

exports[`rectangles determineIgnoreRectangles should return iPhone 12 Pro Max bottom bar rectangles 1`] = `
[
  {
    "bottom": 5,
    "left": 133,
    "right": 148,
    "top": 883,
  },
  {
    "bottom": 5,
    "left": 133,
    "right": 148,
    "top": 883,
  },
]
`;

exports[`rectangles determineIgnoreRectangles should return iPhone 12 bottom bar rectangles 1`] = `
[
  {
    "bottom": 5,
    "left": 133,
    "right": 148,
    "top": 883,
  },
  {
    "bottom": 5,
    "left": 133,
    "right": 148,
    "top": 883,
  },
]
`;

exports[`rectangles determineIgnoreRectangles should return iPhone X bottom bar rectangles 1`] = `
[
  {
    "bottom": 5,
    "left": 133,
    "right": 148,
    "top": 883,
  },
]
`;

exports[`rectangles determineIgnoreRectangles should return nothing if no ignore rectangles need to be returned 1`] = `[]`;

exports[`rectangles determineValidBlockOuts should return the valid blockouts if an array with invalid keys is provided 1`] = `
[
  {
    "bottom": 8,
    "left": 5,
    "right": 7,
    "top": 6,
  },
  {
    "bottom": 14,
    "left": 9,
    "right": 13,
    "top": 12,
  },
]
`;

exports[`rectangles determineValidBlockOuts should return the valid blockouts if an valid array if provided 1`] = `
[
  {
    "bottom": 4,
    "left": 1,
    "right": 3,
    "top": 2,
  },
  {
    "bottom": 4,
    "left": 1,
    "right": 3,
    "top": 2,
  },
]
`;

exports[`rectangles getElementRectangles should return element rectangles 1`] = `
{
  "height": 20,
  "width": 10,
  "x": 1,
  "y": 2,
}
`;
